include/Gaffer/PlugAlgo.h

```diff
@@ -71,6 +71,12 @@ bool canPromote( const Plug *plug, const Plug *parent = NULL );
 /// related to plug layouts are copied to the external
 /// plug - this can be controlled with the `excludeMetadata`
 /// argument.
+///
+/// > Note : When promoting to a Box, this function examines
+/// > the value of the "nodule:type" metadata, and automatically
+/// > creates an intermediate BoxIO node if the plug has a
+/// > nodule. This makes the exact effects of the function dependent
+/// > on whether or not the UI metadata has been loaded.
```

Comment:
Hopefully I've convinced you above that BoxIOs definitely shouldn't disappear from the serialisation. But I'm happy to reconsider our options here for `promote()`, since it is the Achilles heel of this approach. Here's a quick rundown of the possibilities as I see them :

1. The status quo. The downside is a coupling of `promote()` and the "nodule:type" metadata. The upside is that automated scripts can still produce the graph a user would want to see, provided they're run with the metadata loaded, and those scripts don't need to consider the metadata themselves. It's dirty inside MetadataAlgo, but it avoids other code getting its hands dirty.
2. As above, plus moving the metadata to the core modules. Upside is graph creation is deterministic (same whether the UI modules are loaded or not), downside is the core modules are polluted with UI metadata, and we still need the UI files for the widget-specific stuff, so everything UI related isn't all in one place any more. Plus the temptation to start using UI metadata for other purposes.
3. Explicit requests for BoxIO usage in PlugAlgo, with all decisions made by the UI. Upside is we've removed the coupling. Downside is an additional `set<Plug*> plugsThatNeedBoxIO` argument in `Box::create()`, more code in the UI to build that set appropriately (based indirectly on the metadata!), and non-UI-savvy automated scripts never producing BoxIO even though the user would appreciate it. Although this is undoubtedly purer, I'm not convinced it's much of an improvement in real terms, which is why I went with 1.
4. Pure UI sugar. This already failed.
5. BoxIOs appear magically when viewed, disappear when serialised. Ick.
6. A mix of 3 and 5. The UI requests BoxIO nodes explicitly. Automated scripts don't use 'em at all. BoxIO nodes are always serialised. The NodeGraph magically creates BoxIO nodes when first viewing graphs generated by any non-UI-savvy scripts. Upsides are the same as 3, plus a bit of magic. Downsides are getting the magic to work reliably and efficiently (just on first viewing, or when something is promoted naively during viewing?). And what do we do with References? Do we modify them even though they're read-only?
7. What Daniel was really going to suggest, assuming I assumed the wrong thing before.

Of those, I only see 1, 3 and 6 as worthy candidates (unless 7 is awesome). 3 would undoubtedly make me feel less dirty, but I'm not sure that feeling of cleanliness comes with any real-world benefits. 6 is a bit of an unknown, and might be solving a problem we don't really have - I'd at least like to have the need for the magic before implementing it. So from my point of view we're back to the options from the commit comment, perhaps with the additional consideration that 3) does at least prepare the way for 6) should we need it later.